plugin-bsp/src/main/kotlin/org/jetbrains/plugins/bsp/impl/server/connection/DefaultBspConnection.kt

```diff
@@ -120,124 +128,117 @@ private class CancelableInvocationHandlerWithTimeout(
 }
 
 private val log = logger<DefaultBspConnection>()
-private const val CONNECT_SUBTASK_ID = "bsp-file-connection-connect"
+private const val CONNECT_TASK_ID = "bsp-autoconnect"
 
 class DefaultBspConnection(
   private val project: Project,
   private val connectionDetailsProviderExtension: ConnectionDetailsProviderExtension,
 ) : BspConnection {
   private var connectionDetails: BspConnectionDetails? = null
-
   private var server: JoinedBuildServer? = null
   private var capabilities: BazelBuildServerCapabilities? = null
-
   private var bspProcess: Process? = null
-  private var disconnectActions: MutableList<() -> Unit> = mutableListOf()
+  private val disconnectActions: MutableList<() -> Unit> = mutableListOf()
   private val timeoutHandler = TimeoutHandler { Registry.intValue("bsp.request.timeout.seconds").seconds }
-
-  override suspend fun connect(taskId: Any) {
-    try {
-      if (!isConnected()) {
-        doConnect(taskId)
-      }
-    } catch (e: Exception) {
-      bspProcess?.also {
-        if (it.isAlive) it.destroyForcibly()
+  private val mutex = Mutex()
```

Comment:
**suggestion:** Document mutex usage for thread safety

Consider adding a comment explaining that the mutex is used to ensure thread safety and that withLock ensures proper release even in case of exceptions. This will help future maintainers understand the design.